### ISSUE TITLE
transmission-qt crashes on linux

### ISSUE THREAD

**Issue by boreq** [transmission/transmission#5081]
### What is the issue?

Soon after launch the program crashes with the following error:
```
$ transmission-qt
malloc(): corrupted top size
```
Sometimes it crashes with a different error:
```
$ transmission-qt
free(): invalid size
```

```
Process 3023 (transmission-qt) of user 1000 dumped core.

                                                   Stack trace of thread 3030:
                                                   #0  0x00007f017d4a08ec n/a (libc.so.6 + 0x878ec)
                                                   #1  0x00007f017d451ea8 raise (libc.so.6 + 0x38ea8)
                                                   #2  0x00007f017d43b53d abort (libc.so.6 + 0x2253d)
                                                   #3  0x00007f017d43c29e n/a (libc.so.6 + 0x2329e)
                                                   #4  0x00007f017d4aa657 n/a (libc.so.6 + 0x91657)
                                                   #5  0x00007f017d4ade26 n/a (libc.so.6 + 0x94e26)
                                                   #6  0x00007f017d4ae8ed malloc (libc.so.6 + 0x958ed)
                                                   #7  0x00007f017f87383c n/a (libevent-2.1.so.7 + 0xd83c)
                                                   #8  0x00007f017f875b4d evbuffer_add (libevent-2.1.so.7 + 0xfb4d)
                                                   #9  0x000055b7c2e569ff n/a (transmission-qt + 0x1789ff)
                                                   #10 0x000055b7c2e9c92e n/a (transmission-qt + 0x1be92e)
                                                   #11 0x000055b7c2e2cf41 n/a (transmission-qt + 0x14ef41)
                                                   #12 0x00007f017f888c45 n/a (libevent-2.1.so.7 + 0x22c45)
                                                   #13 0x00007f017f88a3af event_base_loop (libevent-2.1.so.7 + 0x243af)
                                                   #14 0x000055b7c2e7e98d n/a (transmission-qt + 0x1a098d)
                                                   #15 0x00007f017d6d72c3 execute_native_thread_routine (libstdc++.so.6 + 0xd72c3)
                                                   #16 0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #17 0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3026:
                                                   #0  0x00007f017d49b766 n/a (libc.so.6 + 0x82766)
                                                   #1  0x00007f017d49e294 pthread_cond_timedwait (libc.so.6 + 0x85294)
                                                   #2  0x00007f017dc8ba0b _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x28ba0b)
                                                   #3  0x00007f017dc86895 n/a (libQt6Core.so.6 + 0x286895)
                                                   #4  0x00007f017dc85a21 n/a (libQt6Core.so.6 + 0x285a21)
                                                   #5  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3023:
                                                   #0  0x00007f017d5139df __poll (libc.so.6 + 0xfa9df)
                                                   #1  0x00007f017d089c2f n/a (libglib-2.0.so.0 + 0xacc2f)
                                                   #2  0x00007f017d0310e2 g_main_context_iteration (libglib-2.0.so.0 + 0x540e2)
                                                   #3  0x00007f017dd4d382 _ZN20QEventDispatcherGlib13processEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE (libQt6Core.so.6 + 0x34d382)
                                                   #4  0x00007f017db35604 _ZN10QEventLoop4execE6QFlagsINS_17ProcessEventsFlagEE (libQt6Core.so.6 + 0x135604)
                                                   #5  0x00007f017db2ce7b _ZN16QCoreApplication4execEv (libQt6Core.so.6 + 0x12ce7b)
                                                   #6  0x000055b7c2d456f6 n/a (transmission-qt + 0x676f6)
                                                   #7  0x00007f017d43c790 n/a (libc.so.6 + 0x23790)
                                                   #8  0x00007f017d43c84a __libc_start_main (libc.so.6 + 0x2384a)
                                                   #9  0x000055b7c2d48845 n/a (transmission-qt + 0x6a845)

                                                   Stack trace of thread 3027:
                                                   #0  0x00007f017d49b766 n/a (libc.so.6 + 0x82766)
                                                   #1  0x00007f017d49e294 pthread_cond_timedwait (libc.so.6 + 0x85294)
                                                   #2  0x00007f017dc8ba0b _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x28ba0b)
                                                   #3  0x00007f017dc86895 n/a (libQt6Core.so.6 + 0x286895)
                                                   #4  0x00007f017dc85a21 n/a (libQt6Core.so.6 + 0x285a21)
                                                   #5  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3024:
                                                   #0  0x00007f017d5139df __poll (libc.so.6 + 0xfa9df)
                                                   #1  0x00007f017c3b726b n/a (libxcb.so.1 + 0xd26b)
                                                   #2  0x00007f017c3b8d1d xcb_wait_for_event (libxcb.so.1 + 0xed1d)
                                                   #3  0x00007f017a269362 n/a (libQt6XcbQpa.so.6 + 0x4e362)
                                                   #4  0x00007f017dc85a21 n/a (libQt6Core.so.6 + 0x285a21)
                                                   #5  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3025:
                                                   #0  0x00007f017d5139df __poll (libc.so.6 + 0xfa9df)
                                                   #1  0x00007f017d089c2f n/a (libglib-2.0.so.0 + 0xacc2f)
                                                   #2  0x00007f017d0310e2 g_main_context_iteration (libglib-2.0.so.0 + 0x540e2)
                                                   #3  0x00007f017dd4d382 _ZN20QEventDispatcherGlib13processEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE (libQt6Core.so.6 + 0x34d382)
                                                   #4  0x00007f017db35604 _ZN10QEventLoop4execE6QFlagsINS_17ProcessEventsFlagEE (libQt6Core.so.6 + 0x135604)
                                                   #5  0x00007f017dc1d778 _ZN7QThread4execEv (libQt6Core.so.6 + 0x21d778)
                                                   #6  0x00007f017e17911f n/a (libQt6DBus.so.6 + 0x2d11f)
                                                   #7  0x00007f017dc85a21 n/a (libQt6Core.so.6 + 0x285a21)
                                                   #8  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #9  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3028:
                                                   #0  0x00007f017d49b766 n/a (libc.so.6 + 0x82766)
                                                   #1  0x00007f017d49e294 pthread_cond_timedwait (libc.so.6 + 0x85294)
                                                   #2  0x00007f017dc8ba0b _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x28ba0b)
                                                   #3  0x00007f017dc86895 n/a (libQt6Core.so.6 + 0x286895)
                                                   #4  0x00007f017dc85a21 n/a (libQt6Core.so.6 + 0x285a21)
                                                   #5  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3031:
                                                   #0  0x00007f017d4e5db5 clock_nanosleep (libc.so.6 + 0xccdb5)
                                                   #1  0x00007f017d4ea987 __nanosleep (libc.so.6 + 0xd1987)
                                                   #2  0x000055b7c2e3941e n/a (transmission-qt + 0x15b41e)
                                                   #3  0x00007f017d6d72c3 execute_native_thread_routine (libstdc++.so.6 + 0xd72c3)
                                                   #4  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #5  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 3029:
                                                   #0  0x00007f017d49b766 n/a (libc.so.6 + 0x82766)
                                                   #1  0x00007f017d49e294 pthread_cond_timedwait (libc.so.6 + 0x85294)
                                                   #2  0x00007f017dc8ba0b _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x28ba0b)
                                                   #3  0x00007f017dc86895 n/a (libQt6Core.so.6 + 0x286895)
                                                   #4  0x00007f017dc85a21 n/a (libQt6Core.so.6 + 0x285a21)
                                                   #5  0x00007f017d49ebb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f017d520d90 n/a (libc.so.6 + 0x107d90)
                                                   ELF object binary architecture: AMD x86-64
```

### Which application of Transmission?

Qt app on Linux, BSD, etc.

### Which version of Transmission?

Transmission 4.0.1 (e1c6e1be43)

**Comment by ckerr** [comment]
@boreq did you build transmission-qt yourself?

**Comment by boreq** [comment]
I have't built it myself, I am using the package that is available on arch linux. That being said I can try building it myself.

**Comment by ckerr** [comment]
I'm using the qt client on a daily basis and haven't seen this issue so there's probably something different either on your system or in the torrents that you're using, so I'd like to get more debugging info out of your setup to figure out what's going on.

Since it's dying in malloc / free, this is probably a memory management issue which is why I'd like you to make an address sanitizer build of transmission-qt:

```sh
$ cd transmission
$ mkdir build
$ cd build
$ cmake -DCMAKE_C_COMPILER="clang" -DCMAKE_CXX_COMPILER="clang++-" -DCMAKE_CXX_FLAGS="-gdwarf-4 -fno-omit-frame-pointer -fsanitize=address,pointer-compare,pointer-subtract,leak,undefined" -DCMAKE_C_FLAGS="-gdwarf-4 -fno-omit-frame-pointer -fsanitize=address,pointer-compare,pointer-subtract,leak,undefined" -DENABLE_QT=ON  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Debug ..
$ make
```

**Comment by ckerr** [comment]
@boreq thank you for building and testing like this; it's very helpful! Could you apply this patch and upload a log of the last 500 lines or so before it dies:

```diff
diff --git a/libtransmission/peer-msgs.cc b/libtransmission/peer-msgs.cc
index 24e3359d6..b5fc935ba 100644
--- a/libtransmission/peer-msgs.cc
+++ b/libtransmission/peer-msgs.cc
@@ -1431,6 +1431,7 @@ ReadState readBtPiece(tr_peerMsgsImpl* msgs, size_t inlen, size_t* setme_piece_b
         auto req = peer_request{};
         msgs->io->read_uint32(&req.index);
         msgs->io->read_uint32(&req.offset);
+        fmt::print("{:s}:{:d} req piece {:d} offset {:d}\n", __FILE__, __LINE__, req.index, req.offset);
         req.length = incoming.length - 9;
         logtrace(msgs, fmt::format(FMT_STRING("got incoming block header {:d}:{:d}->{:d}"), req.index, req.offset, req.length));
         incoming.block_req = req;
@@ -1443,6 +1444,15 @@ ReadState readBtPiece(tr_peerMsgsImpl* msgs, size_t inlen, size_t* setme_piece_b
     auto const block_size = msgs->torrent->blockSize(block);
 
     auto const n_this_pass = std::min(size_t{ req->length }, inlen);
+    auto const* const tor = msgs->torrent;
+    fmt::print("{:s}:{:d} piece size {:d} piece count {:d}\n", __FILE__, __LINE__,  tor->pieceSize(), tor->pieceCount());
+    fmt::print("{:s}:{:d} block size {:d} block count {:d}\n", __FILE__, __LINE__,  tr_block_info::BlockSize, tor->blockCount());
+    fmt::print("{:s}:{:d} loc piece {:d} offset {:d}\n", __FILE__, __LINE__, loc.piece, loc.piece_offset);
+    fmt::print("{:s}:{:d} loc block {:d} offset {:d}\n", __FILE__, __LINE__, loc.block, loc.block_offset);
+    fmt::print("{:s}:{:d} loc byte {:d}\n", __FILE__, __LINE__, loc.byte);
+    fmt::print("{:s}:{:d} loc byte {:d}\n", __FILE__, __LINE__, loc.byte);
+    fmt::print("{:s}:{:d} inlen {:d}\n", __FILE__, __LINE__, inlen);
+    fmt::print("{:s}:{:d} n_this_pass {:d}\n", __FILE__, __LINE__, n_this_pass);
     TR_ASSERT(loc.block_offset + n_this_pass <= block_size);
     if (n_this_pass == 0)
     {
```

**Comment by boreq** [comment]
I think I managed to narrow this down to one specific torrent @ckerr. I paused all other ones. This is the full log for a run with this single torrent unpaused.

[log.txt.gz](https://github.com/transmission/transmission/files/10854169/log.txt.gz)

**Comment by ckerr** [comment]
@boreq I've got a branch open that may fix this assertion failure; it's in the PR linked above and there's a CI-generated tarball at https://github.com/transmission/transmission/suites/11266091938/artifacts/577444778 .

I'm not sure if this assertion failure is related to malloc / free problem or not, or if this is just the entry feel to get to where we can test the malloc / free error :dizzy: so let's see how it goes.

**Comment by boreq** [comment]
> I'm not sure if this assertion failure is related to malloc / free problem or not, or if this is just the entry fee to get to where we can test the malloc / free error dizzy so let's see how it goes.

It is my assumption that the failing assertion led to memory corruption which explains failures with both malloc and free.

**Comment by GASOLINE** [comment]
I have the same issue with Transmission in Fedora xfce. Started after update to version 4.x.
If it crashes, after a minute or so the Fedora crashreport gives me also a link to a general issue?
This is the link: https://retrace.fedoraproject.org/faf/reports/631667/

**Comment by boreq** [comment]
Sorry I didn't have the time to do it earlier.
[log.txt.gz](https://github.com/transmission/transmission/files/10865758/log.txt.gz)

**Comment by ckerr** [comment]
@boreq if this is a public torrent, can you mail it to me at `${firstname}@${firstname}${lastname}.com`?

**Comment by ckerr** [comment]
I've pushed up another another commit to that branch to add some more tracer statements to help figure this out. Please do include the entire log for that single torrent. Thanks!

**Comment by boreq** [comment]
Full log: 
[log.txt.gz](https://github.com/transmission/transmission/files/10866713/log.txt.gz)

**Comment by ckerr** [comment]
@boreq that last log doesn't appear to be from the last push to the branch?

**Comment by boreq** [comment]
```
Your branch and 'origin/fix/5081-assertion-failure-in-readBtPiece' have diverged,
```
Oooops! My bad, I think I got distracted with some Qt-related build problems I had. Here is the fixed log:
[log-5b55ebe99bec1eb62c7abcea24a83989488c6743.txt.gz](https://github.com/transmission/transmission/files/10871683/log-5b55ebe99bec1eb62c7abcea24a83989488c6743.txt.gz)

**Comment by ckerr** [comment]
@boreq 87ccdf38ce2841aabf8a4b714ba2094c3008cde0 is a chonky one. You know what to do :smile_cat:

**Comment by boreq** [comment]
I come with bad news, it seems to consistently crash like this
[log.heap-buffer-overflow.4692621fdf9e77166eb5661358c70d6786e83f58.txt](https://github.com/transmission/transmission/files/10877754/log.heap-buffer-overflow.4692621fdf9e77166eb5661358c70d6786e83f58.txt)
:

**Comment by ckerr** [comment]
You have amazing patience @boreq. Thank you for doing all this testing.

Is this a public torrent you can email to me at my firstname @charleskerr.com?

**Comment by ckerr** [comment]
@boreq I've merged an updated #5097 into `main`. I know it solves _a_ problem but I can't tell if it solves the problem _you're_ seeing. Can you give it another go?

**Comment by boreq** [comment]
This is the result from `main`:

[log.7f9ef4730efbb376243f65761f9dad7b36f3679d.txt](https://github.com/transmission/transmission/files/10890786/log.7f9ef4730efbb376243f65761f9dad7b36f3679d.txt)

The sanitizer reported a use-after-free.

**Comment by ckerr** [comment]
@boreq I don't know how you keep finding these interesting cases; but however you're doing it, I love all the info you're providing.

My own swarms seem so boring in comparison... :smile_cat:

**Comment by ckerr** [comment]
@boreq ^ use-after-free fixed in main

**Comment by boreq** [comment]
I am unable to replicate the crash now. 

I am seeing other issues such as some torrents temporarily not being displayed in the list, oddly slow performance which goes away after restarting transmission qt and torrents needing their data re-verified but no crashes on latest `main`. Those other issues are most likely completely unrelated to the original problem or the fixes.

**Comment by ckerr** [comment]
@boreq there are a couple of other issues tracking those other things you mentioned. But I'm not sure about the torrents needing re-verification; you may want to open a new issue for that one.

Notes: Fixed `4.0.0` crash when receiving malformed piece data from peers.
